Add closest feature.

toregge opened the pull request:
@geirst : please review

toregge committed:
Add closest feature.

geirst commented on searchlib/src/vespa/searchlib/features/closest_feature.cpp:
```diff
@@ -0,0 +1,288 @@
+namespace search::features {
+
+/**
+ * Implements the executor for the closest feature.
```
Consider removing this comment and instead add comments to ClosestSerializedExecutor and ClosestDirectExecutor, making it clear what is the difference between them and when they are used.

geirst commented on searchlib/src/vespa/searchlib/features/closest_feature.cpp:
```diff
@@ -0,0 +1,288 @@
+        return false;
+    }
+    _field_tensor_type = ValueType::from_spec(attr_type_spec);
+    if (_field_tensor_type.is_error() || _field_tensor_type.is_double() || _field_tensor_type.count_mapped_dimensions() == 0) {
```
I suggest being even stricter here: _field_tensor_type.count_mapped_dimensions() != 1 || _field_tensor_type.count_indexed_dimensions() != 1.

geirst commented on searchlib/src/vespa/searchlib/features/closest_feature.cpp:
```diff
@@ -0,0 +1,288 @@
+ClosestDirectExecutor::ClosestDirectExecutor(DistanceCalculatorBundle&& bundle, Value& empty_output, TypedCells identity, const ITensorAttribute& attr)
+    : ClosestExecutor(std::move(bundle), empty_output, identity, attr),
+      _subspace_type(attr.getTensorType()),
+      _labels(attr.getTensorType().count_mapped_dimensions()),
```
This can be simplified with stricter validation in setup(). It will always be 1.

geirst commented on searchlib/src/vespa/searchlib/features/closest_feature.cpp:
```diff
@@ -0,0 +1,288 @@
+ClosestBlueprint::createExecutor(const fef::IQueryEnvironment &env, vespalib::Stash &stash) const
+{
+    auto bundle = _item_label.has_value() ? DistanceCalculatorBundle(env, _field_id, _item_label.value(), "closest") : DistanceCalculatorBundle(env, _field_id, "closest");
+    if (bundle.elements().empty()) {
```
If one of the calculators in the bundle is nullptr, we should also shortcut with an empty tensor.

geirst commented on searchlib/src/vespa/searchlib/features/closest_feature.cpp:
```diff
@@ -0,0 +1,288 @@
+    }
+    _field_tensor_type = ValueType::from_spec(attr_type_spec);
+    if (_field_tensor_type.is_error() || _field_tensor_type.is_double() || _field_tensor_type.count_mapped_dimensions() == 0) {
+        LOG(error, "%s: Field %s invalid type: '%s'", getName().c_str(), _field_name.c_str(), attr_type_spec.c_str());
```
Field %s invalid -> Field %s has invalid ?

geirst reviewed (dismissed):
👍  We should also add unit testing of failing ClosestBlueprint::setup().

toregge committed:
Rename f2 to f when appropriate.
Only allow mixed tensor with 1 sparse and 1 dense dimension.
Return empty tensor if any distance calculator is missing.

toregge committed:
Test closest feature setup failures.

geirst reviewed (approved):
👍
